### DISCUSSION TITLE
Ship Starlight using `.js` + `.d.ts` files

### DISCUSSION THREAD

**Discussion by HiDeoo** [withastro/starlight#983] [Feature Requests]
### What version of `starlight` are you using?

0.11.1

### What is your idea?

Starlight is currently shipped as TypeScript code with no `.d.ts` files. This suggests to compile the TypeScript code and ship `.js` + `.d.ts` files.

### Why is this feature necessary?

This discussion is a follow-up of #861.

Typechecking code importing various parts of Starlight can currently be difficult due to the fact that `@astrojs/starlight` is not compiled and ship TypeScript code with no `.d.ts` files.

For example, an integration importing a TypeScript file from `@astrojs/starlight` could fail to typecheck with `tsc --noEmit` for some stylistic reasons like [`noPropertyAccessFromIndexSignature`](https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature) if the integration uses this rule but `@astrojs/starlight` doesn't as the Starlight shipped TypeScript code will be typechecked with the integration's `tsconfig.json`.

As commented by a TS maintainer in this [comment](https://github.com/microsoft/TypeScript/issues/40426#issuecomment-1373167253) or this [one](https://github.com/microsoft/TypeScript/issues/40426#issuecomment-1199905084), this is expected behavior with such a setup. The solution is to compile the TypeScript code and ship `.js` + `.d.ts` files.

Notes:

- A pre-requisite for this is to have a few remaining TypeScript errors fixed in Starlight (I have already started working on a PR for this).
- Using `astro-scripts` will not be a solution as this is not published

### Participation

- [X] I am willing to submit a pull request for this feature.

**Comment by controversial** [comment]
Is this still planned?

**Reply by HiDeoo** [reply]
Yes, started in https://github.com/withastro/starlight/pull/3572 but paused with the release of Astro v6 and will be picked up relatively soon now that the new Astro major has landed and things stabilize.

**Reply by controversial** [reply]
Subscribed there, thanks!! Looking forward to it & thanks for your hard work

**Reply by filipfalcon** [reply]
0.40 is released but this PR is still an open draft under the v0.40 milestone. What's the current state — is it being retargeted to a later milestone, or blocked on the remaining transpile tasks?

For reference, the underlying problem still reproduces on `@astrojs/starlight@0.40.0`: enabling `noPropertyAccessFromIndexSignature` fails type-checking on shipped source — `user-components/rehype-file-tree.ts:98-99`, `TS4111` (`properties.class`) — which `skipLibCheck` can't suppress since it's `.ts`.

**Reply by HiDeoo** [reply]
Just updated to `v0.42`, as while being pretty closed to finished, `v0.41` will probably only focus on bringing Astro v7 support next week.

**Comment by HiDeoo** [comment]
[Starlight 0.42](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.42.0) now ships JavaScript files with dedicated type declaration files instead of TypeScript source files.
